# Computer Programming

## Programs 2: Getting it Right

These programs build on the previos notebook, and provide more practice.

Remember that we are only concerned with the "Happy Path" at the moment. As you work through these examples, keep in mind that "programs are read much more often than they are written". So, have you chosen good names for your values? Have you left blank lines to show layout? If it comes to that, have you provided prompts that make it obvious what the user should do?

Once you have finished, remember to `add` and `commit` your new file to your Git repository, and then `push` it to GitHub.

## Practice

First here are some questions to check your knowledge. If you don't know, you can Google, ask the person sat next to you, or even ask your friendly tutor. Edit the Markdown as usual (or insert new cells if you prefer - up to you).

_Values in Python should usually be named in "lower snake case". What does this mean? Include some examples._

In Python, “lower snake case” is the recommended style for naming variables, functions, and other identifiers.

What it means
Lower → all letters are lowercase.

Snake case → words are separated by underscores (_), making names easy to read.

This style is part of the PEP 8 guidelines, which define Python’s conventions for clean and readable code.

 Examples of lower snake case
user_name → variable storing a person’s name

total_price → variable holding a calculated price

calculate_area → function name for computing area

max_speed → variable for a car’s maximum speed

is_valid → boolean flag indicating validity

_But some values will be named in "SCREAMING SNAKE CASE". Really. That's what it's called. What does this tell us about the value?_

 What it tells us
The value is not supposed to change during the program’s execution.

It’s a way of communicating to other programmers: “Treat this as fixed.”

Python doesn’t enforce immutability (you technically can reassign it), but by convention, constants are written in SCREAMING SNAKE CASE.

 Examples
PI = 3.14159 → mathematical constant

MAX_SPEED = 120 → maximum speed allowed in a program

DATABASE_URL = "localhost:5432" → configuration constant

DEFAULT_TIMEOUT = 30 → timeout value in seconds

_Jot down three, or four, factors that should be considered when choosing a name for a value._

Clarify : the name should clearly describe what the value represents e.g. "user age" instead of "x".

Consistency: Follow naming conventions like lower snake case for variables and SCREAMING SNAKE CASE for constants.

Length: Keep names long enough to be descriptive but short enough to be practical (e.g., temperature instead of t).

Avoid ambiguity: Don’t use names that could be confusing or misleading (e.g., avoid calling a list of users data).

these factors help code become more readable and maintainable.

_Should you add a space before the bracket after a `print` command? Why?_

you should not add a space before the bracket after a print command in python.

In Python, print is a function, and functions are always called with parentheses immediately after the name.

Writing print ("Hello") with a space is technically allowed (Python will still run it), but it’s considered bad style because it breaks the usual convention.

According to PEP 8 (Python’s style guide), there should be no space between a function name and its opening parenthesis.

_What about spaces around an expression? `2 + 2` or `2+2`?_

the recommended approach is to add spaces around operators in most programming style guides.

it makes to code more readable.

it follows the pep 8 style guide, which is the official python convention.

without spaces, expressions can look cramped and harder to scan quickly.

_What purpose does a blank line serve in a Python program?_

A blank line in a Python program doesn’t affect how the code runs — Python simply ignores it. Its purpose is all about readability and structure.

it is good for separating different sections of code and it improves readablility


_What should be on the last line of every Python program?_

end

_What is a program comment? Why are these almost always best avoided?_

A program comment is text in your code that Python ignores when running the program. It’s written using the # symbol (for single-line comments) or triple quotes """ ... """ for longer explanatory notes. Comments are meant for humans, not the computer — they explain what the code is doing or why.

_Python programmers often refer to `PEP-8`, a much revered document. What is it? (Google is your friend here.)_

PEP 8 is the official Python style guide — a document that sets out the conventions for writing clean, readable, and consistent Python code. It’s one of the most respected references in the Python community and is widely followed across projects.

_Suppose you write a program, and the person sat next to you writes the same program. Assuming both work, would they be exactly the same? Why (not)?_

Even if two people write a program to solve the same problem, the programs are very unlikely to be exactly the same.

they might differ due to things like: naming choices e.g. total and sum, structure, formatting, approach and the order of steps they took.

_AI-generated program code is very easy to spot. Can you work out any features it might have that makes this so?_

Overly neat and consistent formatting: Indentation, spacing, and naming conventions are often perfectly aligned with style guides (like PEP 8), sometimes more than a human would bother with.

Verbose variable names: AI tends to choose long, descriptive names (calculate_total_price) rather than shorter, pragmatic ones (total).

Excessive comments or docstrings: AI often adds comments that explain obvious things (e.g., # add two numbers) which humans usually skip.

Generic structure: Code may look “template‑like” — following textbook examples rather than showing personal quirks or shortcuts.

As usual, feel free to add any other notes here that you might find useful.

## The Programs

Now for the complete programs. Work through these one at a time. Remember you can run them from inside the Notebook.

_Suppose you purchase a 10 inch pizza. How many square *centimetres* of pizza do you have? Write a program below to find out. Remember that 10 inches is the diameter of the pizza._

In [2]:
import math

diameter_inch = 10

radius_inch = diameter_inch / 2

area_inch2 = math.pi * (radius_inch ** 2)

inch_to_cm = 2.54
area_cm2 = area_inch2 * (inch_to_cm ** 2)

print(f"A {diameter_inch}-inch pizza has an area of {area_cm2:.2f} square centimetres.")

A 10-inch pizza has an area of 506.71 square centimetres.


_A 10 inch pizza costs £10.99. A 12 inch pizza costs £13.99. Which is the best buy in terms of pizza area for your pound? Write a program
that prompts for the size and cost, and prints the cost per square cm._

_Hint: It is probably easier to work in pence per square cm. Even then, the answers will be small numbers._

In [6]:
import math

diameter_inch = float(input("Enter the pizza diameter in inches: "))
cost_pounds = float(input("Enter the pizza cost in pounds: "))

cost_pence = cost_pounds * 100

radius_inch = diameter_inch / 2

area_inch2 = math.pi * (radius_inch ** 2)

inch_to_cm = 2.54
area_cm2 = area_inch2 * (inch_to_cm ** 2)

cost_per_cm2 = cost_pence / area_cm2

print(f"A {diameter_inch}-inch pizza costing £{cost_pounds:.2f} "
      f"works out at {cost_per_cm2:.4f} pence per square cm.")

A 10.0-inch pizza costing £10.99 works out at 2.1689 pence per square cm.


A 12.0-inch pizza costing £13.99 works out at 1.9173 pence per square cm.

A 10.0-inch pizza costing £10.99 works out at 2.1689 pence per square cm.

_Write a program where your user enters two integers, which represent weight measurements taken on two different days. Output the difference between the two numbers. The difference can obviously not be a negative number!_

In [7]:

weight_day1 = int(input("Enter the weight on day 1 (in kg): "))
weight_day2 = int(input("Enter the weight on day 2 (in kg): "))

difference = abs(weight_day1 - weight_day2)

print(f"The difference between the two weights is {difference} kg.")

The difference between the two weights is 1 kg.


_The program you just wrote will fail if floating-point numbers are used. This is actually much more likely. Write a version below that
works for floating-point values._

_Suppose the values in your last program were not weights, but the number of sparrows visiting a bird table on two consecutive days. What would you need to change? Write that program below. (You can only have whole numbers of sparrows, obviously.)_

In [10]:

sparrows_day1 = int(input("Enter the number of sparrows on day 1: "))
sparrows_day2 = int(input("Enter the number of sparrows on day 2: "))

difference = abs(sparrows_day1 - sparrows_day2)

print(f"The difference in sparrow visits is {difference}.")

The difference in sparrow visits is 9.


_When packing kiwi fruit, the **usual number** to put in a punnet is four. Write a program that prompts for the number of kiwi fruit to be packed, and displays the number of full punnets that will result._

_Remember that the number of fruits in a punnet might change, so be sure to make that as easy as possible to change._

In [11]:
punnet_size = 4

total_fruit = int(input("Enter the number of kiwi fruit to be packed: "))

full_punnets = total_fruit // punnet_size

print(f"With {total_fruit} kiwi fruit, you can pack {full_punnets} full punnets (of {punnet_size} each).")

With 4 kiwi fruit, you can pack 1 full punnets (of 4 each).


_Peaches, on the other hand, are packed into boxes of 24 fruit. Any remaining after this are packed into smaller punnets of 4. Adapt the program above to handle peaches, and this more complex packing strategy._

In [12]:


box_size = 24
punnet_size = 4

total_peaches = int(input("Enter the number of peaches to be packed: "))

boxes = total_peaches // box_size

remaining_after_boxes = total_peaches % box_size

punnets = remaining_after_boxes // punnet_size

leftover = remaining_after_boxes % punnet_size

print(f"With {total_peaches} peaches, you can pack:")
print(f"- {boxes} full boxes (of {box_size} each)")
print(f"- {punnets} full punnets (of {punnet_size} each)")
print(f"- {leftover} peaches left unpacked")

With 5 peaches, you can pack:
- 0 full boxes (of 24 each)
- 1 full punnets (of 4 each)
- 1 peaches left unpacked


_How about a program to check your working? Write a program that prompts for the number of full boxes and full punnets of peaches, and displays how many peaches this is._

_Hint: Look at your result. Is it as clear as it could be? Never be afraid to write extra code to "spell it out" so that a future programmer can easily understand the code._

In [13]:

box_size = 24
punnet_size = 4

boxes = int(input("Enter the number of full boxes of peaches: "))
punnets = int(input("Enter the number of full punnets of peaches: "))

peaches_in_boxes = boxes * box_size

peaches_in_punnets = punnets * punnet_size

total_peaches = peaches_in_boxes + peaches_in_punnets

print("\nPacking summary:")
print(f"- {boxes} boxes × {box_size} peaches each = {peaches_in_boxes} peaches")
print(f"- {punnets} punnets × {punnet_size} peaches each = {peaches_in_punnets} peaches")
print(f"= Total peaches packed: {total_peaches}")


Packing summary:
- 6 boxes × 24 peaches each = 144 peaches
- 7 punnets × 4 peaches each = 28 peaches
= Total peaches packed: 172


_The final program last week ran like this. For a laugh, write a version below that is as obscure as possible. Break the rules. Make
is **really** difficult to follow._

_A sweet shop sells three types of chocolate: Frodo Bars at 50p, Twux at 75p, and Snookers at 95p. Write a program that asks the user to enter how many of each bar have been purchased, and then prints out the total price._

_The price should be displayed in "pounds and pence". Asking for, say, "150p" makes no sense!_

In [14]:

x,y,z=[int(input(f"{q}?"))for q in("Frodo","Twux","Snookers")]
£=sum([x*50,y*75,z*95]) # total in pence

a,b=divmod(£,100)
print("Total = "+str(a)+" quid and "+str(b)+" shiny pennies")

SyntaxError: invalid character '£' (U+00A3) (2792352379.py, line 2)

## Reflection

Take a look at your last effort, and never, ever do that again!


All the programs in the "library" in your module repo have been formatted using a tool called ``black`` (https://pypi.org/project/black/). Have a read of its docs, and you should see what such things are needed. And take a look at PEP-8 (https://peps.python.org/pep-0008/) too.